pkg/service/one2onerestore/worker_validate.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/scylladb/scylla-manager/v3/pkg/util/parallel"
+	"golang.org/x/sync/errgroup"
+)
+
+// validateClusters checks if 1-1-restore can be performed on provided clusters.
+func (w *worker) validateClusters(ctx context.Context, locations []LocationInfo, nodeMappings []nodeMapping) error {
+	if err := w.client.VerifyNodesAvailability(ctx); err != nil {
+		return errors.Wrap(err, "all nodes must be available")
+	}
+
+	sourceNodeInfo, targetNodeInfo, err := w.collectNodeInfo(ctx, locations)
+	if err != nil {
+		return errors.Wrap(err, "collect nodes info")
+	}
+
+	sourceNodeInfo, err = applyNodeMapping(sourceNodeInfo, getSourceMappings(nodeMappings))
+	if err != nil {
+		return errors.Wrap(err, "apply node mappings")
+	}
+
+	slices.SortFunc(sourceNodeInfo, compareNodesInfo)
+	slices.SortFunc(targetNodeInfo, compareNodesInfo)
+
+	if err := clustersAreEqual(sourceNodeInfo, targetNodeInfo); err != nil {
+		return errors.Wrap(err, "clusters not equal")
+	}
+
+	return nil
+}
+
+func applyNodeMapping(sourceNodeInfo []nodeInfo, sourceMappings map[node]node) ([]nodeInfo, error) {
+	var result []nodeInfo
+	if len(sourceNodeInfo) != len(sourceMappings) {
+		return nil, errors.Errorf("nodes count (%d) != node count in mappings (%d)", len(sourceNodeInfo), len(sourceMappings))
+	}
+	for _, source := range sourceNodeInfo {
+		target, ok := sourceMappings[node{DC: source.DC, Rack: source.Rack, HostID: source.HostID}]
+		if !ok {
+			return nil, errors.Errorf("mapping for source node (%v) is not found", source)
+		}
+		source.DC, source.Rack, source.HostID = target.DC, target.Rack, target.HostID
+		result = append(result, source)
+	}
+	return result, nil
+}
+
+func compareNodesInfo(a, b nodeInfo) int {
+	return cmp.Or(
+		cmp.Compare(a.DC, b.DC),
+		cmp.Compare(a.Rack, b.Rack),
+		cmp.Compare(a.HostID, b.HostID),
+		cmp.Compare(a.CPUCount, b.CPUCount),
+		cmp.Compare(a.StorageSize, b.StorageSize),
+		slices.Compare(a.Tokens, b.Tokens),
+	)
+}
+
+func clustersAreEqual(sourceNodeInfo, targetNodeInfo []nodeInfo) error {
+	if len(sourceNodeInfo) != len(targetNodeInfo) {
+		return errors.Errorf("clusters have different nodes count: source %d != target %d", len(sourceNodeInfo), len(targetNodeInfo))
+	}
+	for i, source := range sourceNodeInfo {
+		target := targetNodeInfo[i]
+		if source.DC != target.DC {
+			return errors.Errorf("source DC doesn't match target DC")
+		}
+		if source.Rack != target.Rack {
+			return errors.Errorf("source Rack doesn't match target Rack")
+		}
+		if source.HostID != target.HostID {
+			return errors.Errorf("source HostID doesn't match target HostID")
+		}
+		if source.CPUCount != target.CPUCount {
+			return errors.Errorf("source CPUCount doesn't match target CPUCount")
+		}
+		if source.StorageSize > target.StorageSize {
+			return errors.Errorf("source StorageSize greater than target StorageSize")
+		}
+		if !slices.Equal(source.Tokens, target.Tokens) {
+			return errors.Errorf("source Tokens doesn't match target Tokens")
+		}
```

Comment:
I'm not sure if we have any guarantee that the tokens are always sorted.
I guess it wouldn't harm to sort them here for extra safety (unless that's ensured by the API or in some other place in the code).